### DISCUSSION TITLE
How to do Dependency Injection with the Prisma Client

### DISCUSSION THREAD

**Discussion by wemersonpdev** [prisma/prisma#13950] [Q&A]
## Problem

I did a lot of research on how to inject the prisma client and I didn't find it, neither to use in my repositories nor in unit tests of my repositories.

## Suggested solution

I developed a way to use dependency injection in prisma which is super elegant and standardized for those who work with dependency injection. I would like this example to be used in the prisma documentation to help others who want to work with prisma dependency injection.

### 1 - Using in Implementation

1.1 - My implementation looks like this:

prisma.implementation.ts
```
import { PrismaClient } from '@prisma/client';

export const prisma: PrismaClient = new PrismaClient({
  errorFormat: 'minimal'
});

```

1.2 - Dependency injection symbol (In this case I use tsyringe for dependency injections, but this works for any other dependency injector.):

prisma.symbol.ts
```
import { container } from 'tsyringe';
import { PrismaClient } from '@prisma/client';

import { prisma } from './implementations/prisma';

container.register<PrismaClient>('PrismaClient', {
  useValue: prisma,
});

````

1.3 - Using in repository:

modelX.repository.ts
```
import { inject, injectable } from 'tsyringe';
import { Prisma, ModelX, PrismaClient } from '@prisma/client';

@injectable()
export class ModelXRepository {
   constructor(
    @inject('PrismaClient')
    private prismaClient: PrismaClient,
   ) {}

   async create(name: string, description: string): Promise<ModelX> {
    const modelXToCreate= {
      name,
      description,
    } as Prisma.ModelXCreateInput;

    const modelXCreated = await this.prismaClient.modelX.create({
      data: modelXToCreate,
    });

    return modelXCreated;
  }
}
```
### 2 - Using in unit tests with Jest

2.1 - Implementation MOCK:

prisma.mock.ts
```
import { PrismaClient } from '@prisma/client';
import { mockDeep, DeepMockProxy } from 'jest-mock-extended';

export type Context = {
  prisma: PrismaClient;
};

export type MockContext = {
  prisma: DeepMockProxy<PrismaClient>;
};

export const createMockContext = (): MockContext => {
  return {
    prisma: mockDeep<PrismaClient>(),
  };
};
```
2.2 - Using in unit test from repository:

modelX.repository.unit.test.ts
```
import 'reflect-metadata';
import { ModelX } from '@prisma/client';

import { Context, createMockContext, MockContext } from '@shared/db/prisma/fakes/prisma.mock';
import { ModelXRepository } from '@modules/modelX/modelX.repository';

describe('Model X Repository tests suites', () => {

  let modelXRepository: ModelXRepository;
  let fakePrismaContext: MockContext;
  
  beforeAll(async () => {
    fakePrismaContext = createMockContext();
    const context = fakePrismaContext as unknown as Context;
  
    modelXRepository = new ModelXRepository(context.prisma);
  });

  it('create: Should create model x', async () => {
    const modelXSuccessDataToCreate = {
      id: 1,
      name: "foo",
      description: "foo bar"
    }
  
    fakePrismaContext.prisma.modelX.create.mockResolvedValue(modelXSuccessDataToCreate );
    
    await expect(modelXRepository.create(modelXSuccessDataToCreate))
      .resolves
      .toEqual(modelXSuccessDataToCreate);
  });
  
  it('create: Not Should create model x because model x with this name exists', async () => {
    const errorUniqueConstraintMessage = 'Unique constraint failed on the fields: (`name`)';

    const modelXExistingToErrorOnCreate = {
      id: 2,
      name: "foo",
      description: "foo bar 2"
    }
  
    fakePrismaContext.prisma.modelX.create.mockRejectedValue(new Error(errorUniqueConstraintMessage));
  
    await expect(modelXRepository.create(modelXExistingToErrorOnCreate))
      .rejects
      .toThrow(new Error(errorUniqueConstraintMessage));
  });
}
```

**Comment by alessandrovaiz-mb** [comment]
Thanks for sharing this solution! Really useful

**Comment by sergiomblabs** [comment]
This is awesome!!  Thanks for share this!

**Comment by ceconcarlsen** [comment]
Thanks for sharing, very helpfull !

**Comment by amorimcode** [comment]
Thanks for sharing this solution! Really useful

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
